Adding support for externally passed random seed and printing used seed on console

cardil opened the pull request:
It's good practice to execute tests in random order. That ensures tests are atomic and not depend on each other or some stateful entity. But, if such error occurs, right now, there is no way to reproduce this exact, erroneous execution.

This PR adds the ability to reproduce those errors by adding support for externally passed random seed. If random seed is not passed, it will be generated automatically. Also, random seed will be printed on console. That's enable it to be achieved by CI server for later.

This operation is strongly influenced by Ruby's rspec order command line option: https://www.relishapp.com/rspec/rspec-core/v/2-13/docs/command-line/order-new-in-rspec-core-2-8

Showcase of execution:

[![asciicast](https://asciinema.org/a/42342.png)](https://asciinema.org/a/42342)

Tibor17 commented:
@cardil Would it be possible without adding a new config parameter random-seed?

cardil commented:
@Tibor17 Yes. Actually, I started development in that way. For example: `-DrunOrder=random:123456`. But, that lead to some dirty code because `RunOrder` is a enum like class, and it's hard to store additional data there. I came up with specific classes to hold randomization and I figure out that it is pointless to make `RunOrder` responsible for two tasks (mode and seed). This is way I decided that additional parameter will be better solution. It's also slightly easier to pass your seed in that way.

cardil commented:
:palm_tree: anyone there?

cardil commented:
Is there something to fix with this PR or everything is okey now?

Tibor17 commented:
Sorry @cardil . I have logged in just now.

Tibor17 commented:
@cardil 
`-DrunOrder=random:123456` I meant as well.
The enums are only some model representation of real world but it does not mean that we cannot change it.
What we should not do is adding a new parameter, especially if we could reuse existing parameter and altering it's behavior and extendig the format of param value. 
We have ambition in Surefire 3 to add parameters but in different way than this approach.
Maybe you was too fast and spent time on this code instead of asking for visions.

joekienzle commented:
We would like this feature as well. Please update this pr

Tibor17 commented:
@joekienzle 
I need to read all this through after 2 years.
@cardil 
Can you meanwhile rebase on master and resolve the git conflicts? Thx.

cardil commented:
Yes. I'll rebase the code.

Tibor17 commented on maven-failsafe-plugin/src/main/java/org/apache/maven/plugin/failsafe/IntegrationTestMojo.java:
```diff
@@ -283,6 +288,22 @@
+     * @since 2.19.2
+     */
+    @Parameter( property = "failsafe.randomSeed", defaultValue = Randomizer.DEFAULT_SEED )
+    private String randomSeed;
```
Can we include this in the already existing parameter `runOrder`?
For instance, the usage can be `-Dfailsafe.runOrder=random:325119`.

Tibor17 commented on maven-surefire-common/src/test/java/org/apache/maven/plugin/surefire/CommonReflectorTest.java:
```diff
@@ -0,0 +1,103 @@
+ * @author <a href="krzysztof.suszynski@wavesoftware.pl">Krzysztof Suszyński</a>
+ * @since 2016-04-03
+ */
+public class CommonReflectorTest extends TestCase
```
Have you registered every new unit test in `JUnit4SuiteTest.java`?
If you have not, please do it and run the build `mvn install`. The integration test may take obe hour to complete.

cardil commented:
I've:
1. merged the code with upstream,
1. change the mechanism to utilize existing runOrder parameter like: `-DrunOrder=random:123456`
1. added some tests to cover the changes
1. fixed some integration tests that ware failing or behave flaky

Please review @Tibor17

cardil commented on surefire-api/src/main/java/org/apache/maven/surefire/util/RunOrders.java:
```diff
@@ -0,0 +1,103 @@
+        if ( !any() )
+        {
+            throw new IllegalStateException(
+                    "20180524:222348 - use #any() method before invoking #firstAsType() method."
```
Oh. It is exception id. A randomly unique generated id that can be easily traced to the place of it's origin. Its only intended for developer bugs, not end user error communication. Its here: https://github.com/wavesoftware/java-eid-exceptions

I use this convention so this slipped. I can remove it.

cardil commented on surefire-its/src/test/resources/testng-parallel-suites/src/test/java/testng/suiteXml/TestNGSuiteTest.java:
```diff
@@ -28,16 +28,23 @@
 
-		TimeUnit.SECONDS.sleep( 2 );
-	}
+        synchronized ( System.out )
```
I've explained it in commit message: https://github.com/apache/maven-surefire/commit/2beb507001867dc61f2be55d206c5825e9f0c2ca

> Fixing flaky test Surefire1177TestngParallelSuitesIT
> 
> This test rely on displaying messages on System.out, but invocations wasn't synchronized so messeges tend to interweave and test wasn't able to assert for them properly. I added synchronization and move code a litle bit.

Tibor17 replied:
@cardil 
Thx for pointing out the synchronization issue with `System.out`.
I have noticed the same problem yesterday, see https://github.com/cbeust/testng/issues/1908
Treating the stream in test would fix our builds but we still have this issue on Ubuntu and not on Windows. Does it mean that the Sysout should be synchronized in the application run? We are setting the delegate via `System.setOut()` and the delegate in `ForkingRunListener` is synchronized. Why we do not see same issue in JUnit tests?

cardil replied:
It depends on how you write to `OutputStream`. Usually you got a single writer, like with using a logging framework. Every thread gets the same instance of writer. Such writer will usually buffer writing to a stream and in that way, each mesage is treated as separate. 

In this test, there are seperate threads that write bytes directly to `OutputStream`, without buffering or synchronization. This is the reason that here synchronization helps. But it might or might not be needed in production code. It depends.

cardil commented:
What's the process? Are we waiting for someone?

eolivelli commented:
We need someone with committer 'karma' to merge the patch.
Like @Tibor17

cardil commented:
Hello, @Tibor17 Yey or Nay?

Tibor17 commented:
@cardil 
Sorry, I was busy this wee. I will have a look in the weekend.

Tibor17 commented:
@cardil 
You have several commits in PR. Can you squash them in one? This simplifies the review, Thx.

Tibor17 commented:
@cardil 
I am currently running all tests in your branch.
Did you also run the build `mvn install -P run-its` and checked that all tests passed?

cardil commented:
Yup. I run them all multiple times. In fact I correct flaky ones. All should pass.

Tibor17 commented:
I hope I am using the right branch:
wavesoftware:feature/random-seed-usage

cardil commented:
Yup. It's the right branch. Those tests are one that I added. They test if running with given seed will produce repeatable build. Ideally it should produce the same outcome on every machine thus those tests I've added. I will dig in to this to determine a fault.

I've manage to reproduce those errors on different machine, with Windows 10. It's very strange, because when I run it directly, everything works properly. This is good news as library works well, only tests are faulty:

```bash
cd surefire-its/target/RunOrderIT_testRandomWithSeed123456
mvn test -Dsurefire.version=3.0.0-SNAPSHOT
```

Running it from IT test fails with errors you just posted. Maybe forked launcher somehow involves?!

I will dig into more and keep you posted.

cardil commented:
Ok. I've manage to fix this bug. Tests are now passing on Linux and Windows on multiple JDK: Azul, Oracle and OpenJDK.

The error was that test randomization was done on list of loaded classes, but class loading doesn't provide a contract on order of loading. So, input list was platform dependent. I've added pre step of sorting input class list. In that way randomization with seed is predictable, every time on every platform.

https://github.com/apache/maven-surefire/pull/112/files#diff-5bf860cde9189b3ffb9b3e0295282663R42

@Tibor17 please review again

cardil commented:
@Tibor17 Please review

Tibor17 commented:
Sorry, I had to work on making stable build with Java 11 and issues which
came up after accepting SUREFIRE-1535. It seems I have got success.

On Wed, Sep 5, 2018 at 11:18 PM Krzysztof Suszyński <
notifications@github.com> wrote:

> @Tibor17 <https://github.com/Tibor17> Please review
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/maven-surefire/pull/112#issuecomment-418884789>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA_yR34zrWUi3qKvf_PNSUJ5k0SXz8A-ks5uYD-agaJpZM4IHvcI>
> .
>


-- 
Cheers
Tibor

seregamorph commented:
This feature looks interesting. @cardil please note, that your PR has conflicting files.

cardil commented:
This code review is taking eons to move in any direction.

If there is someone willing to review and merge this I will fix those conflicts. I hope this will not freeze again.

cardil commented:
Ok. I've rebased the code.

All tests passing. I've executed command:

```mvn clean install -Prun-its -e -V -nsu```

with Maven 3.6.0, and OpenJDK 8u202 running on Ubuntu 18.04

seregamorph commented on pom.xml:
```diff
@@ -514,7 +514,7 @@
         <plugin>
           <artifactId>maven-invoker-plugin</artifactId>
-          <version>3.2.0</version>
+          <version>3.1.0</version>
```
are you sure you intended this?

cardil replied:
Yes. I should have written about it.

I was getting errors when running the `mvn clean install -Prun-its -e -V -nsu` command while building the `maven-failsafe-plugin` subproject. These were NPE errors in the `InstallerTask` class for the `maven-invoker-plugin`. The errors were related to processing the `dependency-reduced-pom.xml` file.

After downgrade to previous version, project build fine.

I was building on OpenJDK 8u202 with Maven 3.6.0 on Ubuntu 18.04

cardil replied:
Update. Today, at work, I've run a same build with invoker `3.2.0` and it installed the file just fine. The error is when you restore build at a point. Like this: `mvn clean install -Prun-its -e -V -nsu -rf maven-failsafe-plugin`

Full error:

```
[INFO] --- maven-shade-plugin:3.1.1:shade (default) @ maven-failsafe-plugin ---
[INFO] Excluding org.apache.maven.surefire:maven-surefire-common:jar:3.0.0-SNAPSHOT from the shaded jar.
[INFO] Excluding org.apache.maven.surefire:surefire-api:jar:3.0.0-SNAPSHOT from the shaded jar.
[INFO] Excluding org.apache.maven.surefire:surefire-logger-api:jar:3.0.0-SNAPSHOT from the shaded jar.
[INFO] Excluding org.apache.maven.surefire:surefire-booter:jar:3.0.0-SNAPSHOT from the shaded jar.
[INFO] Excluding org.apache.maven:maven-toolchain:jar:3.0-alpha-2 from the shaded jar.
[INFO] Excluding org.apache.maven.shared:maven-artifact-transfer:jar:0.11.0 from the shaded jar.
[INFO] Excluding org.apache.maven.shared:maven-common-artifact-filters:jar:3.1.0 from the shaded jar.
[INFO] Excluding commons-codec:commons-codec:jar:1.11 from the shaded jar.
[INFO] Excluding org.slf4j:slf4j-api:jar:1.7.5 from the shaded jar.
[INFO] Excluding org.codehaus.plexus:plexus-java:jar:1.0.3 from the shaded jar.
[INFO] Excluding org.ow2.asm:asm:jar:7.0 from the shaded jar.
[INFO] Excluding com.thoughtworks.qdox:qdox:jar:2.0-M9 from the shaded jar.
[INFO] Excluding org.apache.maven:maven-model:jar:3.0 from the shaded jar.
[INFO] Excluding org.apache.maven:maven-artifact:jar:3.0 from the shaded jar.
[INFO] Excluding org.apache.maven:maven-plugin-api:jar:3.0 from the shaded jar.
[INFO] Excluding org.sonatype.sisu:sisu-inject-plexus:jar:1.4.2 from the shaded jar.
[INFO] Excluding org.sonatype.sisu:sisu-inject-bean:jar:1.4.2 from the shaded jar.
[INFO] Excluding org.sonatype.sisu:sisu-guice:jar:noaop:2.1.7 from the shaded jar.
[INFO] Excluding org.codehaus.plexus:plexus-utils:jar:2.0.4 from the shaded jar.
[INFO] Excluding org.codehaus.plexus:plexus-classworlds:jar:2.2.3 from the shaded jar.
[INFO] Excluding org.codehaus.plexus:plexus-component-annotations:jar:1.7.1 from the shaded jar.
[INFO] Including org.apache.maven.shared:maven-shared-utils:jar:3.1.0 in the shaded jar.
[INFO] Including org.apache.commons:commons-lang3:jar:3.8.1 in the shaded jar.
[INFO] Including commons-io:commons-io:jar:2.6 in the shaded jar.
[INFO] Minimizing jar org.apache.maven.plugins:maven-failsafe-plugin:maven-plugin:3.0.0-SNAPSHOT
[INFO] Minimized 510 -> 105 (20%)
[INFO] Replacing original artifact with shaded artifact.
[INFO] Replacing C:\Users\ksuszynski\git\maven-surefire\maven-failsafe-plugin\target\maven-failsafe-plugin-3.0.0-SNAPSHOT.jar with C:\Users\ksuszynski\git\maven-surefire\maven-failsafe-plugin\target\maven-failsafe-plugin-3.0.0-SNAPSHOT-shaded.jar
[INFO] Dependency-reduced POM written at: C:\Users\ksuszynski\git\maven-surefire\maven-failsafe-plugin\dependency-reduced-pom.xml
[INFO]
[INFO] --- maven-invoker-plugin:3.2.0:install (pre-its) @ maven-failsafe-plugin ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Maven Failsafe Plugin 3.0.0-SNAPSHOT:
[INFO]
[INFO] Maven Failsafe Plugin .............................. FAILURE [ 20.944 s]
[INFO] Maven Surefire Report Plugin ....................... SKIPPED
[INFO] Maven Surefire Integration Tests ................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  21.382 s
[INFO] Finished at: 2019-05-23T11:05:28+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-invoker-plugin:3.2.0:install (pre-its) on project maven-failsafe-plugin: Failed to install project dependencies: MavenProject: org.apache.maven.plugins:maven-failsafe-plugin:3.0.0-SNAPSHOT @ C:\Users\ksuszynski\git\maven-surefire\maven-failsafe-plugin\dependency-reduced-pom.xml: Failed to install project artifacts: null: Failed to install POM: null: NullPointerException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-invoker-plugin:3.2.0:install (pre-its) on project maven-failsafe-plugin: Failed to install project dependencies: MavenProject: org.apache.maven.plugins:maven-failsafe-plugin:3.0.0-SNAPSHOT @ C:\Users\ksuszynski\git\maven-surefire\maven-failsafe-plugin\dependency-reduced-pom.xml
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Failed to install project dependencies: MavenProject: org.apache.maven.plugins:maven-failsafe-plugin:3.0.0-SNAPSHOT @ C:\Users\ksuszynski\git\maven-surefire\maven-failsafe-plugin\dependency-reduced-pom.xml
    at org.apache.maven.plugins.invoker.InstallMojo.installProjectDependencies (InstallMojo.java:468)
    at org.apache.maven.plugins.invoker.InstallMojo.execute (InstallMojo.java:184)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Failed to install project artifacts: null
    at org.apache.maven.plugins.invoker.InstallMojo.installProjectArtifacts (InstallMojo.java:343)
    at org.apache.maven.plugins.invoker.InstallMojo.installProjectDependencies (InstallMojo.java:462)
    at org.apache.maven.plugins.invoker.InstallMojo.execute (InstallMojo.java:184)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Failed to install POM: null
    at org.apache.maven.plugins.invoker.InstallMojo.installProjectPom (InstallMojo.java:401)
    at org.apache.maven.plugins.invoker.InstallMojo.installProjectArtifacts (InstallMojo.java:325)
    at org.apache.maven.plugins.invoker.InstallMojo.installProjectDependencies (InstallMojo.java:462)
    at org.apache.maven.plugins.invoker.InstallMojo.execute (InstallMojo.java:184)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: java.lang.NullPointerException
    at org.apache.maven.plugins.invoker.InstallMojo.installProjectPom (InstallMojo.java:387)
    at org.apache.maven.plugins.invoker.InstallMojo.installProjectArtifacts (InstallMojo.java:325)
    at org.apache.maven.plugins.invoker.InstallMojo.installProjectDependencies (InstallMojo.java:462)
    at org.apache.maven.plugins.invoker.InstallMojo.execute (InstallMojo.java:184)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
```

Tibor17 commented:
@cardil 
Thx for the progress. In #226 we have a tendency not to push users to write a code for surefire nothing but exposing Extensions API which can be implemented by the same user in his POM without asking us in ASF. Nevertheless we should adopt your patch but as an extension which prevents from long waiting periods of our acceptance.

cardil commented:
@Tibor17 Should I do something with this PR? Rewrite it somehow? I do not understand.

cardil commented:
@Tibor17 Should I implement the new extenstion API introduced in https://github.com/apache/maven-surefire/pull/232 or just fix the conflicts?

Tibor17 commented:
@cardil 
Yes please resolve these conflicts. I saw your changes on GitHub already but I have to see them better again in IDE. The runorder is calculated in the forked jvm, so it means the extension won't be possible.

seregamorph commented:
The confusing thing is that if I pass `random:1` it becomes `random:100001`
Using value `random:1` was my first intuitive experiment and initially I thought that my argument was just ignored.
I see that there is a constant `Randomizer.LOWER_BOUND = 100000` limiting it. I do not see reasons not to have value `1` (or even `0`). I changed it and ran tests with seed 1 and 2 and I see that tests are shuffled very well.

cardil commented:
> The confusing thing is that if I pass `random:1` it becomes `random:100001`
> Using value `random:1` was my first intuitive experiment and initially I thought that my argument was just ignored.
> I see that there is a constant `Randomizer.LOWER_BOUND = 100000` limiting it. I do not see reasons not to have value `1` (or even `0`). I changed it and ran tests with seed 1 and 2 and I see that tests are shuffled very well.

My intention on `LOWER_BOUND` was that to ensure that generated seed be always the same length in chars. I think that I will change it to not force values to be changed, so if user gives `random:1` effective seed will be `1`. But if user do not pass a seed, like `random`, then `LOWER_BOUND` will be used to generate new value.

What do you think?

Tibor17 commented:
Can you create a documentation for your feature? Hint, see `maven-surefire-plugin/src/site`.
I am going to review the changes in the weekend but now I have a question to the new constant `RANDOM_SEED` in `BooterConstants`. Was is necessary. Could it be transfered in the value of `randomOrder`?

Tibor17 commented:
Is there related issue in ASF Jira for this PR?

Tibor17 commented:
Try to cover also tests for `AbstractSurefireMojo` and other modified classes. I know it's not easy but I have also spent some time when improving JaCoCo by 1000 lines higher. Do it after rebasing, which is easier then, because now there's several merge conflicts.

Tibor17 commented:
Can you remove this text from JavaDoc from both MOJO classes?
MNG-4979 should be reopened and fixes but it is secondary problem of Maven itself.
```
     * Note that due to bug <a href="https://issues.apache.org/jira/browse/MNG-4979">MNG-4979</a> it's
     * better to configure {@code runOrder} in {@code pom.xml} file using Maven {@code properties},
     * if you feel that you will need to change this parameter from the command line in the future.
     * <br>
     * <br>
```

Tibor17 commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/report/DefaultReporterFactory.java:
```diff
@@ -181,6 +183,30 @@ public void runStarting()
+            final Randomizer randomizer = reportConfiguration.getRunOrderParameters().getRandomizer();
+            final String pluginName = reportConfiguration.getPluginName();
+
+            log( "" );
```
pls remove this line

Tibor17 commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/report/DefaultReporterFactory.java:
```diff
@@ -181,6 +183,30 @@ public void runStarting()
+                            + " with -D%s.runOrder=random:%d",
+                    pluginName, randomizer.getSeed()
+            ) );
+            log( "" );
```
pls remove this line

cardil replied:
This is "\n", just like the code in `runCompleted()` (just beneath) that existed before.

Tibor17 commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/report/DefaultReporterFactory.java:
```diff
@@ -181,6 +183,30 @@ public void runStarting()
+            final String pluginName = reportConfiguration.getPluginName();
+
+            log( "" );
+            log( String.format(
```
pls use debug log here
we use to log config params like this one on debug level
users do not like to see too much logs

cardil replied:
This is crucial place. Without showing the seed to user, by default, this PR have no sense. It must be presented to user be able to reproduce the same exact run.

Tibor17 commented on surefire-api/src/main/java/org/apache/maven/surefire/util/RunOrders.java:
```diff
@@ -0,0 +1,103 @@
+
+    private static List<RunOrderWithArguments> withEmptyArguments( RunOrder[] runOrders )
+    {
+        List<RunOrderWithArguments> orderWithArguments = new ArrayList<RunOrderWithArguments>();
```
pls use diamond `new ArrayList<>();`

seregamorph commented:
@cardil I'd like to appreciate this plugin extension. I've used it for my project to find some test orders that may lead to failures, it's very convenient that I can reproduce exact order on any machine whatever JVM it has. 👍

cardil commented:
> @cardil I'd like to appreciate this plugin extension. I've used it for my project to find some test orders that may lead to failures, it's very convenient that I can reproduce exact order on any machine whatever JVM it has.



> BooterConstants

This property is used to pass a computed random seed to forked process with Booter Serializer/Deserializer. Passing just run order is insufficient. When user do not pass a seed, I generate one and use it to order tests and print it on console.

I can modify runOrder value, bu i feel like that's code smell. Better to leave it in original value and add additional parameter as exact seed - genereted or given.

Example: User gives `-Dsurefire.runOrder=random`. I generate seed, say: `128312` and pass `runOrder => random, seed => 128312` to forked process. When user gives `-Dsurefire.runOrder=random:128312`, I reuse seed given and pass `runOrder => random:128312, seed => 128312` to forked process.

cardil commented:
> Is there related issue in ASF Jira for this PR?

I think not. I've just created this PR 3 years ago. I didn't created an issue.

cardil commented:
> Can you remove this text from JavaDoc from both MOJO classes?
> MNG-4979 should be reopened and fixes but it is secondary problem of Maven itself.
> 
> ```
>      * Note that due to bug <a href="https://issues.apache.org/jira/browse/MNG-4979">MNG-4979</a> it's
>      * better to configure {@code runOrder} in {@code pom.xml} file using Maven {@code properties},
>      * if you feel that you will need to change this parameter from the command line in the future.
>      * <br>
>      * <br>
> ```

Ok, I will remove it. Can you reopen that task? Or create other? I't loooong lasting problem.

cardil commented:
> @cardil I'd like to appreciate this plugin extension. I've used it for my project to find some test orders that may lead to failures, it's very convenient that I can reproduce exact order on any machine whatever JVM it has.

Awesome. I've been using that pattern for years in different languages. It's really helpful. But to use it to full extend we should think of a way of passing this seed to JUnit, TestNG to let those frameworks consume the external seed to order their execution in random, but deterministic way. That would be awesome!

cardil commented:
Ok. I've rebased the code. I will now work on site documentation and on code coverage as @Tibor17 suggested.

Tibor17 commented:
@cardil 
Regarding the comment https://github.com/apache/maven-surefire/pull/112#issuecomment-498011714
I perfectly understand the code smell.
I was thinking about extension abilities by user's SPI. So, if we transfer the string from plugin JVM to the forked JVM means that the use can attach SPI class to the forked JVM and use his own mechanism to 1. parse this string and 2. implement RunOrder. WDYT?

Tibor17 commented on surefire-api/src/main/java/org/apache/maven/surefire/util/RunOrders.java:
```diff
@@ -0,0 +1,103 @@
+{
+    private final List<RunOrderWithArguments> withArguments;
+
+    public RunOrders( RunOrder... runOrders )
```
@cardil 
Why this class exists? Always the is only one runOrder in this array.

Tibor17 commented on surefire-api/src/main/java/org/apache/maven/surefire/testset/DirectoryScannerParameters.java:
```diff
@@ -110,8 +111,15 @@ public boolean isFailIfNoTests()
+        return runOrders;
+    }
+
+    private static RunOrders readRunOrders( String runOrder )
```
This method is never called. Pls remove it.

Tibor17 commented on surefire-api/src/main/java/org/apache/maven/surefire/util/DefaultRunOrderCalculator.java:
```diff
@@ -49,30 +51,33 @@ public DefaultRunOrderCalculator( RunOrderParameters runOrderParameters, int thr
     public TestsToRun orderTestClasses( TestsToRun scannedClasses )
     {
-        List<Class<?>> result = new ArrayList<>( 512 );
+        List<Class<?>> result = new ArrayList<Class<?>>( INITIAL_CAPACITY );
```
Pls make diamond in the constructor as well.

Tibor17 commented:
@cardil 
The `RunOrderCalculator` is able to work with one RunOrder and not more.
I think fetching multiple RunOrders was our problem in ASF.
Redesigning this with max one RunOrder would simplify this PR and our code and the extension too.
Is there any smart reason to accept multiple RunOrders?

Tibor17 commented:
@cardil 
The Apache guys made RunOrder OOP over-complicated. There is unnecessary objects like RunOrder and an array because we are parsing a string of one element to an array with one element, decomposition to object for (which inherently does not support Extensions because of the existence of ala enums) and composition from objects back to a string in `BooterSerializer`.
I think the most simple approach is to use only call the constructor `public RunOrderParameters( String runOrder, File runStatisticsFile )` and not transformin string to `RunOrder`. Then the `RunOrderCalculator` would be as simple as an interpreter of the string and the method `TestsToRun orderTestClasses( TestsToRun scannedClasses )` called in forked JVM. Then we are able to use SPI and the string is transparently moved from Maven process to the forked Surefire process just like that without any complexity.
In your case the calculator would be SPI class with default constructor (one significant change).
Everybody can specify own format and patter except those several default ones.

cardil commented:
@Tibor17 I agree that will significantly simplify the code. But, I'm wondering is there a reason someone addred `RunOrder[]`. Maybe the is some fringe use case for that.

EDIT: I was looking at repo. `RunOrder[]` was introduced here: https://github.com/apache/maven-surefire/commit/13e6736b8cabc04be9b98c71a2a9695391d1f1a1 There is a unit test here: https://github.com/apache/maven-surefire/commit/13e6736b8cabc04be9b98c71a2a9695391d1f1a1#diff-5be1142c69251dcc5baa91176c27d284R42 that showcases the use case for multiple run orders. It's `"failedfirst,alphabetical"`. I think that intention was to order failed first and then alphabetical. It looks a bit like a quick hack to me. In the current code those tests are no more. I wasn't looking at a reason for multiple `RunOrder[]`, before. Maybe failedfirst is now handled in other way?!

t3hk0d3 commented:
Please merge this MR. It is a shame Surefire doesn't have this feature yet, and it is a shame this MR is still open for 4 years.

slawekjaranowski commented:
I hope that this feature was implemented in https://issues.apache.org/jira/browse/SUREFIRE-2055 at long last.

So I close this. 
Thanks everybody for contributions.
